include/ddc/kernels/splines/spline_evaluator.hpp

```diff
@@ -33,37 +46,55 @@ class SplineEvaluator
     {
     };
 
-    using tag_type = typename BSplinesType::tag_type;
+    using tag_type = typename BSplines::tag_type;
 
 public:
+    /// @brief The type of the Kokkos execution space used by this class.
     using exec_space = ExecSpace;
 
+    /// @brief The type of the Kokkos memory space used by this class.
     using memory_space = MemorySpace;
 
-    using bsplines_type = BSplinesType;
-
-    using left_extrapolation_rule_type = LeftExtrapolationRule;
-    using right_extrapolation_rule_type = RightExtrapolationRule;
-
+    /// @brief The type of the interpolation discrete dimension (discrete dimension of interest) used by this class.
     using interpolation_mesh_type = InterpolationMesh;
```

Comment:
This I agree. But then we are going to have different names for the `interpolation_domain_type` in the SplineBuilder and `evaluation_domain_type` in the SplineEvaluator, which looks very confusing to me. This looks like trying hard to "forget" the existence of a symmetry which exists anyway. That's why since the beginning I wanted to stop using the "interpolation" terminology.

I think this is already complicated enough and we should not use two different names for the same thing in the splines.

Reply:
> I think this is already complicated enough and we should not use two different names for the same thing in the splines.

The problem is that these are not at all the same thing. They are only the same thing in the context of a semi-Lagrangian interpolation.

In the context of an interpolation where I want to use a spline approximation to refine my grid they are different.

In the context of a spline solution to a FEM equation they are unrelated.

Reply:
`In the context of a spline solution to a FEM equation they are unrelated` I don't think I agree because for any use-case they are related in the sense that:

SplineBuilder acts on a function defined on X and returns a function defined on BSplineX.
SplineEvaluator acts on a function defined on BSplineX and returns a function defined on X.

Sure the physical units can be different, but this is the same when we do `iFFT(FFT(f)/k^2)`.

Maybe I miss something or this is not what you had in mind ?

Reply:
> SplineBuilder acts on a function defined on X and returns a function defined on BSplineX.
> SplineEvaluator acts on a function defined on BSplineX and returns a function defined on X.

This is true on a continuous domain but if we stick to objects in DDC a discrete definition would be:

> SplineBuilder acts on a function defined on IDimX1 and returns a function defined on BSplineX.
> SplineEvaluator acts on a function defined on BSplineX and returns a function defined on IDimX2.


> this is the same when we do `iFFT(FFT(f)/k^2)`

I think this is explains why I think `SplineEvaluator` and `iFFT` are not equivalent. The inverse Fourier transform is not implemented as an evaluation of the Fourier approximation of the function. The Fourier approximation is defined as:
$f(x) = {\sum_{k=-\infty}}^{\infty} c_k \exp(i2\pi x k/P)$
But the Fourier transform is the (continuous) coefficients $c_n$. The inverse Fourier transform is calculated using:
$f(x_n) = 1/N {\sum_{k=0}}^{N-1} c_k \exp(i 2\pi n k/N)$
Because $x_n$ doesn't appear in the RHS you can't use the iFFT to evaluate the function at an arbitrary point.

For splines we have something similar in that:
- the builder is equivalent to `FFT`
    - Builder : $c_k = (B^-1f(x_j))_k$
    - FFT : $c_k = {\sum_{j=0}}^{N-1} f(x_j) \exp(i 2\pi k j/N)$
- the spline coefficients are equivalent to the Fourier transforms
     - Builder : $c_k$
     - FFT : $c_k$
- the evaluator is equivalent to evaluating the approximation
     - Spline Evaluator : $f(x) = {\sum_{i=0}}^N c_n b_i(x)$
     - Fourier Evaluator : $f(x) = {\sum_{k=-\infty}}^{\infty} c_k \exp(i2\pi x k/P)$
- Multiplying the matrix from the builder by a vector of spline coefficients is equivalent to `iFFT`. 
     - Spline inverse transform : $f(x_j) = (B c_i)_j
     - iFFT : $f(x_j) = 1/N {\sum_{k=0}}^{N-1} c_k \exp(i 2\pi j k/N)

It would be possible to implement an inverse Spline Builder just as it would be possible to implement a Fourier evaluator but this is not commonly done because these 2 tools are rarely used to solve the same problem.

In the majority of cases Fourier is used to manipulate values at fixed points. This is useful for (amongst others):
- derivatives (thanks to the properties of Fourier space)
- noise removal (by eliminating coefficients that correspond to highly oscillating modes)

It would be possible to use a spline to calculate a derivative using a Builder(iBuilder(f)) method. From the coefficients $c_{k,d}$ of a d-th degree spline representation you can obtain the coefficients $c'_{k,d-1}$ of the (d-1)-th spline representation of the derivative using $d_x b_{i,d}(x) = d (b_{i,d-1}(x)/(k_{i+d}-k_i) - b_{i+1, d-1}(x)/(k_{i+d+1}-k_{i+1}))$ You would then have to find the coefficients $c'_{k,d}$ somehow. As you can see this is significantly more complicated than just evaluating the function.

Suppressing a signal does make sense using splines. You can set one of the coefficients to 0 to damp the function in that region, but it is not very useful.

This is why we don't usually apply the matrix in an inverse Spline Builder.

On the spline side, spline approximations are usually used to approximate functions. This is useful for (amongst other things):
- interpolation to determine values of functions at points where they are not already known
- creating smooth functions which minimise certain properties (e.g. splines are often the solution to energy minimisation problems)

For interpolation we use an "interpolating spline" which "interpolates" a function f(x) at known "interpolation points":
![image](https://github.com/CExA-project/ddc/assets/25363473/9e867035-abb7-4e37-aef3-366fe21326eb)

The interpolating spline (in red) is used to approximate the function at arbitrary points x. It is possible to do this with a Fourier approximation by evaluating the equation above, however while a spline approximation will tend to be smoother than the original function, the Fourier approximation will tend to be noisier:
![image](https://github.com/CExA-project/ddc/assets/25363473/c9d160bb-4e0c-4300-ae28-168c8a6b2966)

This is why we don't usually use a Fourier approximation to interpolate a function.

Reply:
I guess I reached the limit for the maximum number of equations in a comment :laughing:

Reply:
> I think this is already complicated enough and we should not use two different names for the same thing in the splines.

I think it is significantly more complicated to use the same name for two similar but slightly different concepts than to use 2 names. Having different names makes it easier to explain the difference clearly.

The mathematical convention is:
![image](https://github.com/CExA-project/ddc/assets/25363473/b0e0b21a-c3c4-4f1f-94e7-e26347c806e2)

You could argue that every point of every function is an interpolation point as every function interpolates itself but I think this is just proof that the concept of "interpolation point" is not relevant for every function

Reply:
#435

Reply:
Yes I would be happy to use the term `interpolation points` for an FFT. But if you are aiming for a common API I would recommend using terminology from the splines as it is used in more contexts. The Fourier terminology tends to be quite specific to Fourier. But if for example you added a Lagrange interpolator then the builder/evaluator would contain the same concepts

Reply:
Thx sorry you were right from the beginning, As an evaluation can contain less or more information (depending on the "arrival" mesh) than the spline coefficients this is not an (inverse) transformation. This is the case in semi-Lagrangian though but this is not the general case.

Should I "`sed -i s/interpolation/evaluation/g`" the SplineEvaluator and SplineEvaluator2D in another branch ?